contracts/rollover/CrossCurrencyRollover.sol

```diff
@@ -0,0 +1,415 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.18;
+
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "@openzeppelin/contracts/token/ERC721/utils/ERC721Holder.sol";
+
+import "@uniswap/v3-periphery/contracts/interfaces/ISwapRouter.sol";
+
+import "../origination/OriginationController.sol";
+
+import "../interfaces/ICrossCurrencyRollover.sol";
+import "../interfaces/ILoanCore.sol";
+import "../interfaces/IRepaymentController.sol";
+import "../libraries/LoanLibrary.sol";
+import "../libraries/InterestCalculator.sol";
+
+import {
+    CCR_StateAlreadySet,
+    CCR_Paused,
+    CCR_InvalidState,
+    CCR_SideMismatch,
+    CCR_SameCurrency,
+    CCR_CollateralMismatch,
+    CCR_LenderIsBorrower,
+    CCR_CallerNotBorrower,
+    CCR_ZeroAddress
+} from "../errors/Rollover.sol";
+
+contract CrossCurrencyRollover is ICrossCurrencyRollover, OriginationController, ERC721Holder {
```

Comment:
I apologize for only noticing this now, but I see we are doing a _full inherit_ of OriginationController - this would mean that this contract could start regular loans too, not only cross-currency rollovers. I don't think this is ideal, since it eliminates the security/safety benefit from having rollovers in a separate contract from normal loan origination. If we give `ORIGINATOR_ROLE` to this contract, we are giving it the power to do _all_ forms of origination this contract supports, and right now this contract supports normal loan origination (not just rollovers).

I see that these are the functions you "need" from `OriginationController`:

- `isApproved`
- `isApprovedForContract`
- `_recoverSignature`

Maybe it's better to move all the functions in OC under `PERMISSION MANAGEMENT` and `SIGNATURE VERIFICATION` to a new contract, something like `OriginationControllerBase`? Then both the original OC and `CrossCurrencyRollover` can both inherit the base contract, but the "core origination" and "CC rollover" functionality is separate and not in the same contract (the same fix is also needed for `OriginationControllerMigrate`, but that can be done separately from CC rollovers). Then you can also separate roles such that `ROLLOVER_MANAGER_ROLE` isn't in the `OriginationController` (where it's not needed).